### ISSUE TITLE
OS X errors while compiling from source

### ISSUE THREAD

**Issue by mattbasta** [facebook/hhvm#3765]
In attempting to install HHVM locally from source, I received the following errors. What's inline below are what seem to be the interesting bits. The full terminal output is here: https://gist.github.com/mattbasta/6c30883fd797af52bdac

```
...
Scanning dependencies of target hphp_parser
[ 16%] Building CXX object third-party/folly/CMakeFiles/folly.dir/folly/GroupVarint.cpp.o
[ 16%] Building CXX object third-party/thrift/CMakeFiles/hphp_thrift.dir/lib/cpp/async/TAsyncTimeout.cpp.o
[ 16%] Building CXX object hphp/parser/CMakeFiles/hphp_parser.dir/hphp.tab.cpp.o
hphp.y:11:41: fatal error: hphp/compiler/parser/parser.h: No such file or directory
 #include "hphp/compiler/parser/parser.h"
                                         ^
compilation terminated.
make[2]: *** [hphp/parser/CMakeFiles/hphp_parser.dir/hphp.tab.cpp.o] Error 1
make[1]: *** [hphp/parser/CMakeFiles/hphp_parser.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 16%] Building CXX object third-party/folly/CMakeFiles/folly.dir/folly/IPAddress.cpp.o
[ 16%] Building CXX object third-party/thrift/CMakeFiles/hphp_thrift.dir/lib/cpp/async/TAsyncTimeoutSet.cpp.o
[ 16%] Building CXX object third-party/thrift/CMakeFiles/hphp_thrift.dir/lib/cpp/async/TEventBase.cpp.o
[ 16%] Building CXX object third-party/folly/CMakeFiles/folly.dir/folly/IPAddressV4.cpp.o
[ 17%] Building CXX object third-party/thrift/CMakeFiles/hphp_thrift.dir/lib/cpp/async/TEventBaseManager.cpp.o
Linking C static library libsqlite3.a

...

[ 19%] Building CXX object third-party/thrift/CMakeFiles/hphp_thrift.dir/lib/cpp/util/FdUtils.cpp.o
[ 19%] Building CXX object third-party/folly/CMakeFiles/folly.dir/folly/Uri.cpp.o
[ 19%] Building CXX object third-party/folly/CMakeFiles/folly.dir/folly/Version.cpp.o
Linking CXX static library libhphp_thrift.a
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: libhphp_thrift.a(TTransportException.cpp.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: libhphp_thrift.a(TTransportException.cpp.o) has no symbols
[ 19%] Built target hphp_thrift
[ 19%] Building CXX object third-party/folly/CMakeFiles/folly.dir/folly/detail/CacheLocality.cpp.o
[ 19%] Building CXX object third-party/folly/CMakeFiles/folly.dir/folly/detail/FunctionalExcept.cpp.o
[ 20%] Building CXX object third-party/folly/CMakeFiles/folly.dir/folly/detail/MallocImpl.cpp.o
[ 20%] Building CXX object third-party/folly/CMakeFiles/folly.dir/folly/experimental/io/FsUtil.cpp.o

...

[ 21%] Building CXX object third-party/folly/CMakeFiles/folly.dir/gen/EscapeTables.cpp.o
[ 21%] Building CXX object third-party/folly/CMakeFiles/folly.dir/gen/FingerprintTables.cpp.o
[ 21%] Building CXX object third-party/folly/CMakeFiles/folly.dir/gen/FormatTables.cpp.o
[ 21%] Building CXX object third-party/folly/CMakeFiles/folly.dir/gen/GroupVarintTables.cpp.o
Linking CXX static library libfolly.a
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: libfolly.a(MallocImpl.cpp.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: libfolly.a(InlineExecutor.cpp.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: libfolly.a(MallocImpl.cpp.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: libfolly.a(InlineExecutor.cpp.o) has no symbols
[ 21%] Built target folly
make: *** [all] Error 2
```

I'm on OS X 10.9.3 with XCode 5.1.1. I did attempt to uninstall and reinstall Boost per the 10.9 compile instructions, but it did not help.

**Comment by wjywbs** [comment]
@sgolemon 
I used gcc 4.9.2 and also had same ld error when building hhvm 3.6.0 on mac 10.10.
Xcode's clang had this bug: https://llvm.org/bugs/show_bug.cgi?id=22106
Then I used clang 3.5 or 3.6 with libc++, but there were errors in the libc++ headers like
```
thrift/lib/cpp/concurrency/FunctionRunner.h:83:6: error: 
      no matching constructor for initialization of 'VoidFunc'
      (aka 'function<void ()>')
   : func_(std::bind(func, arg)), repFunc_(0), initFunc_(0)
/usr/local/Cellar/llvm36/3.6.0/lib/llvm-3.6/bin/../include/c++/v1/functional:1448:5: note: 
      candidate constructor not viable......
```
Finally I built hhvm using clang 3.6 with gcc's c++ headers, but only `hhvm --version` works. It crashed in Sweepable when accessing _thread variable. `-Wl,-segaddr,__text,0` did not link or run properly in mac 10.10 because the pagezero section must be at address 0. `-Wl,-pagezero_size,0x00001000` worked.
